internal/nfs/nodeserver/nodeserver.go

```diff
@@ -0,0 +1,346 @@
+/*
+Copyright 2022 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nodeserver
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	csicommon "github.com/ceph/ceph-csi/internal/csi-common"
+	"github.com/ceph/ceph-csi/internal/util"
+	"github.com/ceph/ceph-csi/internal/util/log"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	mount "k8s.io/mount-utils"
+	netutil "k8s.io/utils/net"
+)
+
+const (
+	defaultMountPermission = uint64(0o777)
+	// Address of the NFS server.
+	paramServer           = "server"
+	paramShare            = "share"
+	paramClusterID        = "clusterID"
+	mountPermissionsField = "mountPermissions"
+)
+
+// NodeServer struct of ceph CSI driver with supported methods of CSI
+// node server spec.
+type NodeServer struct {
+	csicommon.DefaultNodeServer
+}
+
+// NewNodeServer initialize a node server for ceph CSI driver.
+func NewNodeServer(
+	d *csicommon.CSIDriver,
+	t string,
+) *NodeServer {
+	return &NodeServer{
+		DefaultNodeServer: *csicommon.NewDefaultNodeServer(d, t, map[string]string{}),
+	}
+}
+
+// NodePublishVolume mount the volume.
+func (ns *NodeServer) NodePublishVolume(
+	ctx context.Context,
+	req *csi.NodePublishVolumeRequest,
+) (*csi.NodePublishVolumeResponse, error) {
+	err := validateNodePublishVoluemRequest(req)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	// Considering kubelet make sure the stage and publish operations
+	// are serialized, we dont need any extra locking in nodePublish.
+
+	volumeID := req.GetVolumeId()
+	volCap := req.GetVolumeCapability()
+	targetPath := req.GetTargetPath()
+	mountPermissions := defaultMountPermission
+	mountOptions := volCap.GetMount().GetMountFlags()
+	if req.GetReadonly() {
+		mountOptions = append(mountOptions, "ro")
+	}
+
+	var server, baseDir, clusterID string
+	for k, v := range req.GetVolumeContext() {
```

Comment:
some of these parameters are not optional, there needs to be a better verification. Ideally use helper functions for getting the name of the server, the nfs-export (share is Windows/Samba terminology) and others. Return a clear failure when a required option is missing (also easier to add unit-tests for helper functions)

Reply:
This is done valiateNodePublishRequest()

Reply:
oh, right! That feels a little hidden, I prefer to have a check where the value is actually fetched. Multiple helper functions make it simpler, I think. Not a **must**, but would be _nice_ to have the error checking a little more obvious.